hazelcast/src/test/java/com/hazelcast/map/ExpirationListenerTest.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.core.EntryEvent;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.IMap;
+import com.hazelcast.map.listener.EntryEvictedListener;
+import com.hazelcast.map.listener.EntryExpiredListener;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.TestHazelcastInstanceFactory;
+import com.hazelcast.test.annotation.ParallelTest;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.util.Random;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelTest.class})
+public class ExpirationListenerTest extends HazelcastTestSupport {
+
+    private static final int instanceCount = 3;
+    private static final Random rand = new Random();
+
+    private static HazelcastInstance[] instances;
+    private IMap map;
+
+    @Before
+    public void init() {
+        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(instanceCount);
+        Config config = new Config();
+        instances = factory.newInstances(config);
+        HazelcastInstance node = getInstance();
+        map = node.getMap(randomMapName());
+    }
+
+    private HazelcastInstance getInstance() {
+        return instances[rand.nextInt(instanceCount)];
+    }
+
+
+    @Test
+    public void testExpirationListener_notified_afterExpirationOfEntries() throws Exception {
+        int numberOfPutOperations = 1000;
+        CountDownLatch expirationEventArrivalCount = new CountDownLatch(numberOfPutOperations);
+
+        map.addEntryListener(new ExpirationListener(expirationEventArrivalCount), true);
+
+        for (int i = 0; i < numberOfPutOperations; i++) {
+            map.put(i, i, 100, TimeUnit.MILLISECONDS);
+        }
+
+        // wait expiration of entries.
+        sleepSeconds(1);
```

Comment:
Could we get rid of the sleep at all? Maybe with assertTrueEventually? With the upcoming parallel running tests I worry about more and more racy tests otherwise.